Error while compiling bitcoin with qt support

The error details are:

    In file included from qt/clientmodel.cpp:5:0:
    qt/clientmodel.h: In function ‘void BlockTipChanged(ClientModel*, bool, const CBlockIndex*)’:
    qt/clientmodel.h:92:10: error: ‘void ClientModel::numBlocksChanged(int, const QDateTime&, double)’ is protected
         void numBlocksChanged(int count, const QDateTime& blockDate, double nVerificationProgress);
              ^
    qt/clientmodel.cpp:256:154: error: within this context
             Q_EMIT clientmodel->numBlocksChanged(pIndex->nHeight, QDateTime::fromTime_t(pIndex->GetBlockTime()), clientmodel->getVerificationProgress(pIndex));
                                                                                                                                                          ^

It looks like a static function is trying to call a private method of ClientModel. There are two ways to fix this problem:

1. Declare numBlocksChanged as a friend of class ClientModel.
2. Add another public method to call line 256 instead of putting it in a static method outside class. 

I tried two approaches and both of them work. Which one is better?

Btw, I am new here. I am trying to join bitcoin community and hope I can make more contributions.

--- a/src/qt/clientmodel.cpp
+++ b/src/qt/clientmodel.cpp
@@ -253,7 +253,10 @@ static void BlockTipChanged(ClientModel *clientmodel, bool initialSync, const CB
     // if we are in-sync, update the UI regardless of last update time
     if (!initialSync || now - nLastBlockTipUpdateNotification > MODEL_UPDATE_DELAY) {
         //pass a async signal to the UI thread
-        Q_EMIT clientmodel->numBlocksChanged(pIndex->nHeight, QDateTime::fromTime_t(pIndex->GetBlockTime()), clientmodel->getVerificationProgress(pIndex));
+        QMetaObject::invokeMethod(clientmodel, "numBlocksChanged", Qt::QueuedConnection,
+                                  Q_ARG(int, pIndex->nHeight),
+                                  Q_ARG(QDateTime, QDateTime::fromTime_t(pIndex->GetBlockTime())),
+                                  Q_ARG(double, clientmodel->getVerificationProgress(pIndex)));
         nLastBlockTipUpdateNotification = now;
     }
 }

